python/graphstorm/graphpartitioning/metadata_schema.py

```diff
@@ -0,0 +1,799 @@
+                                # node type ntype1 are located.
+                            }
+                        feature2: {
+                            ...
+                            },
+                        ...
+                        },
+                    ntype2: {
+                        ...
+                        },
+                    ...
+
+                }
+            * ``edge_data``[optional], a dictionary which contains the metadata
+                    for the features associated with edge types from the graph.
+                    This dictionary is identical to the ``node_data``
+                    dicionary, but it will describe the metadata which will
+                    be used to locate files for edge features for edge types
+                    from the input graph.
+
+    A valid metadata.json file should adhere to the following rules.
+        * ``graph_name`` should be a valid string with non-zero length,
+        * ``num_nodes_per_type`` should be a list of positive non-zero integers
+                and the length of list should be more than 0,
+        * ``num_edges_per_type`` should be a list of positive non-zero integers
+                and the length of list should be more than 0,
+        * ``node_type`` should ba list of valid non-None strings of positive
+                length, length of this list should be equal to the
+                ``num_nodes_per_type`` list,
+        * ``edge_type`` should ba list of valid non-None strings of positive
+                length, length of this list should be equal to the
+                ``num_edges_per_type`` list,
+        * ``edges`` should have the same no. of keys are the no. of edge types
+                present in the ``edge_type`` list,
+        * ``node_data``, if present, can contain keys whose count is less than
+                or equal to the node types present in the ``node_type`` list,
+                Each node type in this dictionary is expected to non-zero
+                features,
+        * ``edge_data``, if present, follows the same rules as described for
+                ``node_data`` in the earlier bullet point.
+
+    Examples:
+    ---------
+    Creating and initalizaing the singleton instance during the pipelines
+    initialization process.
+    >>> obj = Metadata()
+    >>> base_dir = "/home/ubuntu/graph-dataset"
+    >>> file_path = "metadata.json"
+    >>> obj.init(file_path, base_dir)
+
+    After initialization is completed, general usage of any method or property
+    in this class is
+    >>> ntypes = Metadata().ntypes # To retrieve node types from the graph
+    >>> # To get the id <-> ntype map
+    >>> # id_ntype_map will be {0: ``ntype0``, 1: ``ntype1``, ...}
+    >>> id_ntype_map = Metadata().id_ntype_map
+    """
+
+    _instance = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(MetadataSchema, cls).__new__(cls)
+        return cls._instance
```

Comment:
As discussed previously: I don't understand the need for a singleton design for this class.

What's wrong with having `my_metadata = MetadataSchema(config_file_path)`? A singleton object has _very_ specific use cases, this seems like over-engineering.

Reply:
Following is the reasoning behind adapting singleton pattern for the use case in this context. 
1. Need to maintain only one instance of the metadata object throughout the application with its reference being stored in one single place. 
2. Build the metadata object once and use it everywhere it is needed, without having to build it from scratch multiple times or passing around reference to this object or storing copies of this objects reference in multiple places. One of my requirements is the need to NOT exceed more than 5 function parameters to any of the function in the entire pipeline. For this purpose singleton pattern could be an ideal pattern. All the user has to do is to get the object's reference either using class constructor call (as implemented currently) or use a class method to get the only instance of this class. 
3. Metadata information is used almost everywhere in the pipeline. All the user has to do is to get the reference to this object and use the object appropriately as needed. 
4. Information flow from this metadata class to the pipelines application logic is only in one direction and it is read-only access to the user who is using this object. 

`my_metadata` = MetadataSchema(config_file_path)`. The only issue with this code is that the object will be rebuilt wherever this piece of code is used. And all the associated logic will be executed every single time this line is used in the application. 

We have to agree on an option which adds most value to the application.

Reply:
> Need to maintain only one instance of the metadata object throughout the application with its reference being stored in one single place.

Why not instantiate an object and pass it as a function argument wherever it's needed then?

> One of my requirements is the need to NOT exceed more than 5 function parameters to any of the function in the entire pipeline.

As mentioned previously, the 5 argument rule is a code style _suggestion_ I don't think we should modify our design just to try to stick to this rule. Where it makes sense to have more than 5 arguments it's OK, as long as we try to keep it reasonable (15 arguments is definitely too many, 6-7 can be OK).

> Metadata information is used almost everywhere in the pipeline. All the user has to do is to get the reference to this object and use the object appropriately as needed.

The can do that without a singleton pattern though?

> The only issue with this code is that the object will be rebuilt wherever this piece of code is used.

The idea is that you initialize the object once then use it wherever necessary. There's no need to re-instantiate it.

Reply:
The way I understand your comments is that create an instance of the metadata schema object and pass it around where ever it is needed. 

Please note that **all the modules use MetadataSchema object, expect 1 or 2 modules.** 

What this means is that literally every function in this application will have metadata schema object as a parameter !!!, Doesn't this look odd ? Doesn't this call usage of some pattern (singleton, factory etc...) ?. 

If you still insist on creating metadata object and passing it around, I will do so. Please give it a thought and confirm. This change will trigger a lot of code changes on my end and consequently re-coding and testing time on it.

Reply:
Only the constructor of each class that needs the metadata schema needs to take it as an argument, not every function.

```python
class MetadataSchema():
    def __init__(self, config_file):
        # Init here

class MyOtherClassThatNeedsMetadata():
    def __init__(self, metatadata_schema: MetadataSchema):
        self.metadata_schema = metatadata_schema
    def method_that_needs_metadata(self):
        # We don't need to pass metadata again to this or any class method
        do_something_with_metadata(self.metadata_schema)
```

So at most this would require the constructor of the class that needs the metadata to take it as an argument.

How is the singleton pattern more appropriate than this?

Reply:
Resolving in favor of regular class object.